File: SimulationJava/src/Swiat.java
import java.io.File;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.util.Formatter;
import java.util.Random;
import java.util.Scanner;

public abstract class Swiat{
    static private int Losowanie =40;
    static private int wylosowanoWilka =1;
    static private int wylosowanoOwce =2;
    static private int wylosowanoLisa =3;
    static private int wylosowanoZolwia =4;
    static private int wylosowanoAntylope =5;
    static private int wylosowanoCyberOwce =6;
    static private int wylosowanoTrawe =7;
    static private int wylosowanoMlecza =8;
    static private int wylosowanoGuarane =9;
    static private int wylosowanoJagody =10;
    static private int wylosowanoBarszcz =11;
    protected int xMaks;
    protected int yMaks;
    protected int ileS;
    protected int ileSB;
    protected int wymPolaX;
    protected int wymPolaY;
    private int gdzieCz;
    private int czasUmiejetnosci=5;
    protected Organizm[][] tab;
    private Organizm head=null;
    private boolean umiejetnosc = false;
    private boolean czlowiekZyje = false;
    private String log;
    Organizm czlowiek;

    //private String log;
    //private int tura;

  

    public abstract int[][] getSasiad(Organizm o);
    public abstract int[][] getSasiadBarszczu(Organizm o);

    public int getWymPolaX() {
        return wymPolaX;
    }

    public int getWymPolaY() {
        return wymPolaY;
    }

    public int getIleS() {
        return ileS;
    }

    public int getIleSB() {
        return ileSB;
    }

    public int getCzasUmiejetnosci() {
        return czasUmiejetnosci;
    }

    public void setLog(String log) {
        this.log = log;
    }

    public String getLog() {
        return log;
    }

    public boolean getUmiejetnosc(){
        return umiejetnosc;
    }

    public void setUmiejetnosc(boolean umiejetnosc) {
        this.umiejetnosc = umiejetnosc;
    }

    public boolean getCzlowiekZyje(){
        return czlowiekZyje;
    }

    public int getGdzieCz() {
        return gdzieCz;
    }

    public void setGdzieCz(int gdzieCz) {
        this.gdzieCz = gdzieCz;
    }

    public void setCzlowiekZyje(boolean czlowiekZyje) {
        this.czlowiekZyje = czlowiekZyje;
    }

    public void setCzasUmiejetnosci(int czasUmiejetnosci) {
        this.czasUmiejetnosci = czasUmiejetnosci;
    }
    
    public void setCzlowiek(Organizm czlowiek) {
        this.czlowiek = czlowiek;
    }

    public void generujZapis(){
        String z="";
        Organizm obecny =head;
        while(obecny!=null){
            if(obecny.getZycie()!=0){
                z += obecny.wypiszNazwe()+","+obecny.getX()+","+obecny.getY()+","+obecny.getSila()+","+obecny.getPlodnosc();
                if(obecny==czlowiek) z+=","+czasUmiejetnosci+","+umiejetnosc;
                z+="\n";
            }
            obecny=obecny.getNas();
        }
        File zapis = new File("save.txt");
        if(!zapis.exists())
            try {
                zapis.createNewFile();
            } catch (IOException e) {
                e.printStackTrace();
            }
        if(zapis.canWrite()){
             Formatter fm;
            try {
                fm = new Formatter(zapis);
             fm.format("%s", z);
             fm.close();
             System.out.println("wygenerowano zapis 1");
            } catch (FileNotFoundException e) {
                e.printStackTrace();
            }
        }
        z=xMaks+","+yMaks;
        if(this instanceof Hexagony) z+=",H";
        else z+=",K";
        File plansza = new File("swiat.txt");
        if(!plansza.exists())
             try {
                plansza.createNewFile();  
             } catch (IOException e) {
                 e.printStackTrace();
              }
        if(plansza.canWrite()){
            Formatter fm;
            try {
                fm = new Formatter(plansza);
                fm.format("%s", z);
                fm.close();
                System.out.println("wygenerowano zapis 2");
            } catch (FileNotFoundException e) {
                e.printStackTrace();
            }
        }
    }

    public void wczytaj(){
        int x,y,sila,plodnosc,g;
        String xx,yy,silaa,plodnoscc,czass,um,napis,zwierze;
        try {
            Scanner odczyt = new Scanner(new File("save.txt"));
            while(odczyt.hasNextLine()){
                napis=odczyt.nextLine();
                zwierze="";
                g=0;
                xx="";
                yy="";
                silaa="";
                plodnoscc="";
                czass="";
                um="";
                for(int i=0;i<napis.length();i++){
                    if(napis.charAt(i)==',') g++;
                    else{
                        if(g==0) zwierze+=napis.charAt(i);
                        if(g==1) xx+=napis.charAt(i);
                        if(g==2) yy+=napis.charAt(i);
                        if(g==3) silaa+=napis.charAt(i);
                        if(g==4) plodnoscc+=napis.charAt(i);
                        if(g==5) czass+=napis.charAt(i);
                        if(g==6) um+=napis.charAt(i);
                    }
                }
                x=Integer.parseInt(xx);
                y=Integer.parseInt(yy);
                sila=Integer.parseInt(silaa);
                plodnosc=Integer.parseInt(plodnoscc);
                Organizm n=null;
                if("czlowiek".equals(zwierze)){
                    czasUmiejetnosci=Integer.parseInt(czass);
                    if("true".equals(um)) umiejetnosc=true;
                    n=new Czlowiek(x, y, this);
                    this.czlowiek=n;
                    czlowiekZyje=true;
                }
                if("wilk".equals(zwierze)) n=new Wilk(x, y, this);
                if("antylopa".equals(zwierze)) n=new Antylopa(x, y, this);
                if("lis".equals(zwierze)) n=new Lis(x, y, this);
                if("zolw".equals(zwierze)) n=new Zolw(x, y, this);
                if("cyber-owca".equals(zwierze)) n=new CyberOwca(x, y, this);
                if("owca".equals(zwierze)) n=new Owca(x, y, this);
                if("trawa".equals(zwierze)) n=new Trawa(x, y, this);
                if("mlecz".equals(zwierze)) n=new Mlecz(x, y, this);
                if("barszcz sosnowskiego".equals(zwierze)) n=new Barszcz(x, y, this);
                if("guarana".equals(zwierze)) n=new Guarana(x, y, this);
                if("wilcze jagody".equals(zwierze)) n=new Jagody(x, y, this);
                if(n!=null){
                    dodaj(x,y,n);
                    n.setPlodnosc(plodnosc);
                    n.setSila(sila);
                }
            }
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        }
    }

    public void dodaj(int i,int j, Organizm n){
        tab[i][j]=n;
        if(head==null){
            head=n;
        } else {
            Organizm obecny = head;
            if(obecny.getInicjatywa() < n.getInicjatywa()){
                n.setNas(obecny);
                obecny.setPop(n);
                head=n;
            } else {
                while(obecny.getNas()!=null){
                    if(obecny.getNas().getInicjatywa() < n.getInicjatywa()) break;
                    obecny=obecny.getNas();
                }
                n.setPop(obecny);
                if(obecny.getNas()!=null){
                    n.setNas(obecny.getNas());
                    obecny.getNas().setPop(n);
                }
                obecny.setNas(n);
            }
        }
    }

    protected void ustawianie(){
        int czlowiekX,czlowiekY;
        Random rand = new Random();
        czlowiekX = rand.nextInt(xMaks);
        czlowiekY = rand.nextInt(yMaks);
        for (int i = 0; i < xMaks; i++) {
            for (int j = 0; j < yMaks; j++) {
                int x = rand.nextInt(Losowanie);
                if(i==czlowiekX && j==czlowiekY){
                    dodaj(i,j,new Czlowiek(i, j,this));
                    czlowiekZyje = true;
                    czlowiek=getPole(i, j);
                    continue;
                }
                if(x==wylosowanoWilka) dodaj(i,j,new Wilk(i, j,this));
                if(x==wylosowanoOwce) dodaj(i,j,new Owca(i, j,this));
                if(x==wylosowanoLisa) dodaj(i,j,new Lis(i, j,this));
                if(x==wylosowanoZolwia) dodaj(i,j,new Zolw(i, j,this));
                if(x==wylosowanoAntylope) dodaj(i,j,new Antylopa(i, j,this));
                if(x==wylosowanoCyberOwce) dodaj(i,j,new CyberOwca(i, j,this));
                if(x==wylosowanoTrawe) dodaj(i,j,new Trawa(i, j,this));
                if(x==wylosowanoMlecza) dodaj(i,j,new Mlecz(i, j,this));
                if(x==wylosowanoGuarane) dodaj(i,j,new Guarana(i, j,this));
                if(x==wylosowanoJagody) dodaj(i,j,new Jagody(i, j,this));
                if(x==wylosowanoBarszcz) dodaj(i,j,new Barszcz(i, j,this));
            }
        }
    }


    
   
    private void organizmy(){
        Organizm obecny=head;
        while(obecny!=null){
            Organizm nas = obecny.getNas();
            if (obecny.plodnosc != 0)obecny.setPlodnosc(obecny.getPlodnosc()-1);;
            if(obecny.getZycie()==1){
                obecny.akcja();
            } 
            else {
                if(obecny.getZycie() == 0){
                    if(obecny.wypiszNazwe()=="czlowiek") czlowiekZyje = false;
                    obecny.smierc();
                }
            }

            obecny=nas;
        }
        if(czlowiekZyje)if(czlowiek.getZycie()==0)czlowiekZyje=false;
    }

    private void start(){
        Organizm obecny = head;
	while (obecny != null) {
		if(obecny.zycie != 0 )obecny.setZycie(1);
		else {
            if(obecny.wypiszNazwe()=="czlowiek") czlowiekZyje = false;
            obecny.smierc();
		}
		obecny = obecny.getNas();
	}
    }

    public void symuluj(){
        if(umiejetnosc){
            czasUmiejetnosci--;
            if(czasUmiejetnosci==0) umiejetnosc=false;
        } else if(czasUmiejetnosci!=5) czasUmiejetnosci++;
        this.log="";
        start();
        organizmy();
        gdzieCz=10;
    }

    public Organizm getPole(int xx,int yy){
        return tab[xx][yy];
    }


    public int getXM(){
        return this.xMaks;
    }

    public int getYM(){
        return this.yMaks;
    }

    public void setTab( int x,int y, Organizm n){
        this.tab[x][y]=n;
    }

    public void setHead(Organizm head) {
        this.head = head;
    }


}

File: SimulationCpp/projektPO/swiat.cpp
#include"swiat.h"
#include"wilk.h"
#include"owca.h"
#include"lis.h"
#include"zolw.h"
#include"antylopa.h"
#include"mlecz.h"
#include"trawa.h"
#include"guarana.h"
#include"jagody.h"
#include"barszcz.h"
#include"cyberOwca.h"
#include"czlowiek.h"
#include <iostream>
#include <cstdlib>
#include <cstdio>
#include <fstream>
#include <conio.h>
using namespace std;

Swiat::Swiat(int a, int b):xMaks(a),yMaks(b) {
	tab = new Organizm**[a];
	for (int i = 0; i < a; i++) tab[i] = new Organizm*[b];

	for (int i = 0; i < a; i++) {
		for (int j = 0; j < b; j++) {
			tab[i][j] = nullptr;
		}
	}
}
//dodawanie do listy kolejnych organizmow tak aby zostaly ulozone zgodnie z inicjatywa i wiekiem
//element zostaje dodany albo na koniec listy albo przed organizmem o mniejszej inicjatywie
void Swiat::dodaj(int i, int j, Organizm* n) {
	tab[i][j] = n;
	if (!Head) {
		Head = n;
	}
	else {
		Organizm* obecny = Head;
		if (obecny->getInicjatywa() < n->getInicjatywa()) {
			n->setNas(obecny);
			obecny->setPop(n);
			Head = n;
		}
		else {
			while (obecny->getNas() != nullptr ) {
				if (obecny->getNas()->getInicjatywa() < n->getInicjatywa()) break;
				obecny = obecny->getNas();
			}
			n->setPop(obecny);
			if (obecny->getNas()) {
				n->setNas(obecny->getNas());
				obecny->getNas()->setPop(n);
			}
			obecny->setNas(n);
		}
	}
}
//losowanie na kazde miejsce organizmu
//czlowiek jest losowany wczesniej i ustawiany gdy indeksy tablicy sie zgadzaja (moze byc tylko jeden czlowiek)
void Swiat::ustawianie() {
	int czlowiekX, czlowiekY;
	czlowiekX = rand() % xMaks;
	czlowiekY = rand() % yMaks;
	for (int i = 0; i < xMaks; i++) {
		for (int j = 0; j < yMaks; j++) {
			if (tab[i][j] == nullptr) {
				int x = rand() % SZANSA;
				if (i == czlowiekX && j == czlowiekY) x = MIEJSCECZLOWIEKA;
				if (x == WYLOSOWANOWILKA) {
					Organizm* wilk = new Wilk(i, j, this);
					dodaj(i, j,wilk);
				}
				if (x == WYLOSOWANOOWCE) {
					Organizm* owca = new Owca(i, j, this);
					dodaj(i, j, owca);
				}
				if (x== WYLOSOWANOLISA) {
					Organizm* lis = new Lis(i, j,this);
					dodaj(i, j, lis);
				}
				if (x == WYLOSOWANOZOLWIA) {
					Organizm* zolw = new Zolw(i, j, this);
					dodaj(i, j, zolw);
				}
				if (x == WYLOSOWANOANTYLOPE) {
					Organizm* antylopa = new Antylopa(i, j, this);
					dodaj(i, j, antylopa);
				}	
				if (x == WYLOSOWANOMLECZA) {
					Organizm* mlecz = new Mlecz(i, j, this);
					dodaj(i, j, mlecz);
				}
				if (x == WYLOSOWANOTRAWE) {
					Organizm* trawa = new Trawa(i, j, this);
					dodaj(i, j, trawa);
				}
				if (x == WYLOSOWANOGUARANE) {
					Organizm* guarana = new Guarana(i, j, this);
					dodaj(i, j, guarana);
				}
				if (x == WYLOSOWANOWILCZEJAGODY) {
					Organizm* jagody = new Jagody(i, j, this);
					dodaj(i, j, jagody);
				}
				if (x == WYLOSOWANOBARSZCZ) {
					Organizm* barszcz = new Barszcz(i, j, this);
					dodaj(i, j, barszcz);
				}
				if (x == WYLOSOWANOCYBEROWCE) {
					Organizm* cyberOwca = new CyberOwca(i, j, this);
					dodaj(i, j, cyberOwca);
				}
				if (x == MIEJSCECZLOWIEKA && !czlowiekZyje) {
					Organizm* czlowiek = new Czlowiek(i, j, this);
					dodaj(i, j, czlowiek);
					czlowiekZyje = true;
				}
			}
		}
	}
}

void Swiat::generujZapis() {
	fstream plik,plik2;
	plik.open("zapis.txt", ios::trunc);
	plik.close();
	plik.open("zapis.txt", ios::out);
	plik2.open("zapis-wielkosc.txt", ios::trunc);
	plik2.close();
	plik2.open("zapis-wielkosc.txt", ios::out);
	plik2 << xMaks << "," << yMaks;
	plik2.close();
	Organizm* obecny = Head;
	plik << tura << endl;
	while (obecny) {
		if(obecny->zycie!=0)plik << obecny->wypiszNazwe()<<","<<obecny->getX() << "," << obecny->getY() << "," << obecny->zycie << "," << obecny->getSila() << ",";
		if (obecny->wypiszNazwe() == "czlowiek")plik << umiejetnosc << "," << czasUmiejetnosci<<",";
		plik << endl;
		obecny = obecny->getNas();
	}
	plik.close();
}

void Swiat::wczytajZapis() {
	fstream plik;
	plik.open("zapis.txt", ios::in);
	int coSieZapisuje=10;
	string napis;
	string orgX="", orgY="", nazwa="",zycie="", sila = "",tu="", umiejetnoscs="", turyUmiejetnosci="";
	while (getline(plik, napis)) {
		orgX = "";
		orgY = "";
		nazwa = "";
		zycie = "";
		sila = "";
		umiejetnoscs = "";
		turyUmiejetnosci = "";
		for (int i = 0; i < napis.length(); i++) {
			if (napis[i] == ',') coSieZapisuje++;
			else {
				if (coSieZapisuje == 0) nazwa += napis[i];
				if (coSieZapisuje == 1) orgX += napis[i];
				if (coSieZapisuje == 2) orgY += napis[i];
				if (coSieZapisuje == 3) zycie += napis[i];
				if (coSieZapisuje == 4) sila += napis[i];
				if (coSieZapisuje == 5) umiejetnoscs += napis[i];
				if (coSieZapisuje == 6) turyUmiejetnosci += napis[i];
				if (coSieZapisuje == 10) tu += napis[i];
			}
		}
		if (coSieZapisuje == 10) {
			coSieZapisuje = 0;
			continue;
		}
		coSieZapisuje = 0;
		int ox= atoi(orgX.c_str()), oy= atoi(orgY.c_str()), oz= atoi(zycie.c_str()), os= atoi(sila.c_str());
		Organizm* n = nullptr;
		if (nazwa == "wilk") n = new Wilk(ox,oy,this);
		if (nazwa == "owca") n = new Owca(ox, oy, this);
		if (nazwa == "lis") n = new Lis(ox, oy, this);
		if (nazwa == "antylopa") n = new Antylopa(ox, oy, this);
		if (nazwa == "czlowiek") {
			int ile = 0;
			n = new Czlowiek(ox, oy, this);
			czlowiekZyje = true;
			if (umiejetnoscs == "1") {
				umiejetnosc = true;
				ile = 1;
			}
			else ile = -1;
			czasUmiejetnosci = atoi(turyUmiejetnosci.c_str()) + ile;
		}
		if (nazwa == "cyber-owca") n = new CyberOwca(ox, oy, this);
		if (nazwa == "zolw") n = new Zolw(ox, oy, this);
		if (nazwa == "trawa") n = new Trawa(ox, oy, this);
		if (nazwa == "mlecz") n = new Mlecz(ox, oy, this);
		if (nazwa == "guarana") n = new Guarana(ox, oy, this);
		if (nazwa == "barszcz sosnowskiego") n = new Barszcz(ox, oy, this);
		if (nazwa == "wilcze jagody") n = new Jagody(ox, oy, this);
		if (n) {
			n->setSila(os);
			n->zycie = oz;
			cout << n->wypiszNazwe() << ": x:" << n->getX() << ", y: " << n->getY() << endl;
			dodaj(ox, oy, n);
		}
	}
	tura = atoi(tu.c_str())-1;
}



void Swiat::rysuj()const {
	system("cls");
	cout << "Kacper Puda s188625 gr.4" << endl << endl;
	char podloga = SYMBOLPODLOGI,krawedz= SYMBOLKRAWEDZI,bok= SYMBOLBOKU;
	cout << "  "<<bok;
	for (int j = 0; j < yMaks; j++) cout << j / 10;
	cout << bok << "  ";
	cout << endl;
	cout << "  " << bok;
	for (int j = 0; j < yMaks; j++) cout << j % 10;
	cout << bok << tura;
	cout << endl;
	cout << podloga<<podloga<<krawedz;
	for (int j = 0; j < yMaks; j++) cout << podloga;
	cout << krawedz << podloga << podloga ;
	cout << endl;
	for (int i = 0; i < xMaks; i++) {
		cout << i / 10 << i % 10 << bok;
		for (int j = 0; j < yMaks; j++) {
			SetConsoleTextAttribute(GetStdHandle(STD_OUTPUT_HANDLE),170);
			if (tab[i][j] != nullptr) tab[i][j]->wypisz();
			else cout << " ";
		}
		SetConsoleTextAttribute(GetStdHandle(STD_OUTPUT_HANDLE), FOREGROUND_GREEN | FOREGROUND_BLUE | FOREGROUND_RED);
		cout << bok<< i / 10 << i % 10 << endl;
	}
	cout << podloga << podloga << krawedz;
	for (int j = 0; j < yMaks; j++) cout << podloga;
	cout << krawedz << podloga << podloga;
	cout << endl;
	cout << "  " << bok;
	for (int j = 0; j < yMaks; j++) cout << j / 10;
	cout << bok << "  ";
	cout << endl;
	cout << "  " << bok;
	for (int j = 0; j < yMaks; j++) cout << j % 10;
	cout << bok << "  ";
	cout << endl;
}
//wykonywanie akcji zwierzat ktore zyja i nie sa siwerzo narodzone oraz usuwanie martwych zwierzat
void Swiat::organizmy() {
	Organizm* obecny = Head;
	while (obecny) {
		Organizm* nas =obecny->getNas();
		if (obecny->plodnosc != PLODNY)obecny->plodnosc--;
		if (obecny->zycie != MARTWY && obecny->zycie != MLODY)obecny->akcja();
		else {
			if (obecny->zycie == MARTWY) {
				obecny->smierc(Head);
				obecny = nullptr;
			}
		}
		obecny = nas;
	}
}
//zmienianie swiezo narodzonych organizmow w zywe
void Swiat::start() {
	Organizm* obecny = Head;
	while (obecny) {
		if(obecny->zycie != MARTWY)obecny->zycie = ZYWY;
		else {
			if (obecny->wypiszNazwe() == "czlowiek") czlowiekZyje = false;
		}
		obecny = obecny->getNas();
	}
}

void Swiat::coRobiCzlowiek(char &znak) {
	char gc1 = STRZALKA, gc2=NIC;
	int i = PIERWSZAAKCAJA;
	if (umiejetnosc) {
		czasUmiejetnosci--;
		cout << "Umiejetnosc czlowieka aktywna. Ilosc pozotalych tur: " << czasUmiejetnosci << endl;
	}
	else {
		if (czasUmiejetnosci < 5) {
			czasUmiejetnosci++;
			if(czasUmiejetnosci !=  5)cout << "Ladowanie umiejetnosci. Ilosc pozotalych tur: " << 5 - czasUmiejetnosci << endl;
		}
		if (czasUmiejetnosci == 5) cout << "Umiejetnosc moze zostac aktywowana" << endl;
	}
	if (czasUmiejetnosci == 0) umiejetnosc = false;
	cout << "Co robi czlowiek:";
	gdzieCzlowiek = STOI;
			while ((gc1 != NASTEPNATURA && gc1 != KONIEC && gc1 != ZAPIS) || (gc2 != NASTEPNATURA && gc2 != KONIEC && gc2 != 'z')) {
				gc1 = _getch();
				if (gc1 == NASTEPNATURA || gc1 == KONIEC || gc1 == ZAPIS) {
					znak = gc1;
					break;
				}
				else {
					if (gc1 == UMIEJETNOSCCZ) {
						if (czasUmiejetnosci == 5 && !umiejetnosc) {
							umiejetnosc = true;
							cout << " aktywuje umiejetnosc";
							i++;
						}
					}
					if (gc1 != STRZALKA) continue;
				}
				gc2 = _getch();
				if (gc2 == NASTEPNATURA || gc2 == KONIEC) {
					znak = gc1;
					break;
				}
				if (i != 0) cout << ",";
				if (gc2 == STRZALKAWDOL) {
					cout << " idzie w dol ";
					gdzieCzlowiek = DOL;
				}
				if (gc2 == STRZALKAWGORE) {
					cout << " idzie w gore";
					gdzieCzlowiek = GORA;
				}
				if (gc2 == STRZALKAWLEWO) {
					cout << " idzie w lewo";
					gdzieCzlowiek = LEWO;
				}
				if (gc2 == STRZALKAWPRAWO) {
					cout << " idzie w prawo";
					gdzieCzlowiek = PRAWO;
				}
				if (i == PIERWSZAAKCAJA) i++;
			}
			cout << endl;
}

void Swiat::symuluj() {
	while (true) {
		char znak = NIC;
		tura++;
		start();
		cout << log << endl;
			if (czlowiekZyje) {
				coRobiCzlowiek(znak);
			}
			
				while (znak != KONIEC && znak != NASTEPNATURA) {
					
					if (znak == ZAPIS) {
						cout << "generowanie zapisu" << endl;
						generujZapis();
					}
					znak = _getch();
				}
				if (znak == KONIEC) break;
			
		
		log = "";
		this->organizmy();
		this->rysuj();
	}
}

int Swiat::getXM() const {
	return this->xMaks;
}

int Swiat::getYM()const {
	return this->yMaks;
}

Swiat::~Swiat() {
	for (int i = 0; i < xMaks; i++) {
		for (int j = 0; j < yMaks; j++) {
			if (tab[i][j]) {
				delete[]tab[i][j];
			}
		}
	}
	cout << "Koniec symulacji!" << endl;
}